### PR TITLE
yp-WOR-455 Update listWorkspaces

### PR DESCRIPTION
**Pull request by ghost**
https://broadworkbench.atlassian.net/browse/WOR-455

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -765,12 +765,38 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+        // be reachable when running against the dev environment.
+        val reachableWorkspaces = workspaces.filter { workspace =>
+          workspace.workspaceType match {
+            case WorkspaceType.McWorkspace =>
+              try {
+                workspaceManagerDAO.getWorkspace(workspace.workspaceIdAsUUID, userInfo.accessToken)
+                true
+              } catch {
+                case e: ApiException =>
+                  logger.warn(s"MC Workspace ${workspace.name} (${workspace.workspaceIdAsUUID}) does not exist in the current WSM instance. ")
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:779]: I think the only ApiException that can be reasonably caught here are 404s, correct? In that case this should check the error code on the exception

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -765,12 +765,38 @@ class WorkspaceService(protected val userInfo: UserInfo,
...
+                  false
+              }
+            case _ => true
+          }
+        }
+
+          reachableWorkspaces.map { workspace =>
+            val cloudPlatform = workspace.workspaceType match {
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:787]: It looks like we're making a second call to `getWorkspace` here, is that redundant given the one above on L775?

**Review by aherbst-broad — COMMENTED** [review]
Looking good, had a few questions.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 2 above)
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:787]: It's not redundant unless we store the first result from `getWorkspace` or do some sort of combination filtering/mapping. I couldn't figure out how to do the exception handling with a `collect`, but I thought Scala experts might have some ideas.

On the other hand, I would expect that the `getWorkspace` call is O(n), and thus calling it twice shouldn't introduce much inefficiency. @aherbst-broad am I incorrect about the cost of calling `getWorkspace`?

**Comment by cahrens** [comment]
@yunip20 your branch is showing commits that have already merged. You can try to rebase, or you could simply cherrypick your commits on top of the branch I started (or start a new branch from develop… just make sure to pull from develop before you add your changes so it is up-to-date).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:779]: Good point about checking the error code.

**Pull request closed without merging**